File: src/rpc/protocol.ts
import type { RpcCommand } from "./types.js";

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const nonEmpty = (value: unknown, field: string): string => {
  if (typeof value !== "string" || value.trim() === "") throw new Error(`invalid command: ${field} must be a non-empty string`);
  return value;
};

const optionalId = (record: Record<string, unknown>): { id?: string } => {
  if (record.id === undefined) return {};
  return { id: nonEmpty(record.id, "id") };
};

const positiveLimit = (value: unknown): number | undefined => {
  if (value === undefined) return undefined;
  if (!Number.isInteger(value) || (value as number) < 1 || (value as number) > 100) {
    throw new Error("invalid command: limit must be an integer between 1 and 100");
  }
  return value as number;
};

const positiveInteger = (value: unknown, field: string): number | undefined => {
  if (value === undefined) return undefined;
  if (!Number.isSafeInteger(value) || (value as number) < 1) {
    throw new Error(`invalid command: ${field} must be a positive integer`);
  }
  return value as number;
};

const nonNegativeCursor = (value: unknown): number | undefined => {
  if (value === undefined) return undefined;
  if (!Number.isInteger(value) || (value as number) < 0) {
    throw new Error("invalid command: cursor must be a non-negative integer");
  }
  return value as number;
};

const stringArray = (value: unknown, field: string): readonly string[] => {
  if (!Array.isArray(value) || value.some(item => typeof item !== "string")) {
    throw new Error(`invalid command: ${field} must be an array of strings`);
  }
  return value;
};

const patch = (value: unknown, field = "patch"): Record<string, unknown> => {
  if (!isRecord(value)) throw new Error(`invalid command: ${field} must be an object`);
  return value;
};

/** Runtime parser shared by legacy and v1 so malformed input never reaches a handler. */
export const parseRpcCommand = (value: unknown): RpcCommand => {
  if (!isRecord(value) || typeof value.type !== "string" || value.type === "") {
    throw new Error("invalid command: missing type field");
  }
  const base = optionalId(value);
  const type = value.type;
  switch (type) {
    case "initialize": {
      if (!Number.isInteger(value.version)) throw new Error("invalid command: version must be an integer");
      if (value.clientName !== undefined && typeof value.clientName !== "string") throw new Error("invalid command: clientName must be a string");
      return { ...base, type, version: value.version as number, ...(value.clientName === undefined ? {} : { clientName: value.clientName }) };
    }
    case "prompt": case "steer": case "follow_up":
      return { ...base, type, message: nonEmpty(value.message, "message") };
    case "abort": case "get_state": case "get_messages": case "get_available_models": case "compact":
    case "session_list": case "session_save": case "config_get": case "mcp_list": case "skills_list":
    case "dream_run": case "instruction_files_list":
      return { ...base, type };
    case "set_model": return { ...base, type, modelId: nonEmpty(value.modelId, "modelId") };
    case "set_auto_compaction": {
      if (typeof value.enabled !== "boolean") throw new Error("invalid command: enabled must be a boolean");
      return { ...base, type, enabled: value.enabled };
    }
    case "approval_response": {
      if (typeof value.approved !== "boolean") throw new Error("invalid command: approved must be a boolean");
      return { ...base, type, requestId: nonEmpty(value.requestId, "requestId"), approved: value.approved };
    }
    case "clarification_response": return { ...base, type, requestId: nonEmpty(value.requestId, "requestId"), answer: nonEmpty(value.answer, "answer") };
    case "session_new": return { ...base, type, ...(value.modelId === undefined ? {} : { modelId: nonEmpty(value.modelId, "modelId") }) };
    case "session_load": {
      if (value.includeMessages !== undefined && typeof value.includeMessages !== "boolean") {
        throw new Error("invalid command: includeMessages must be a boolean");
      }
      return {
        ...base,
        type,
        sessionId: nonEmpty(value.sessionId, "sessionId"),
        ...(value.includeMessages === undefined ? {} : { includeMessages: value.includeMessages }),
      };
    }
    case "session_branch": {
      if (!Number.isInteger(value.messageId) || (value.messageId as number) < 1) throw new Error("invalid command: messageId must be a positive integer");
      if (value.summarize !== undefined && typeof value.summarize !== "boolean") throw new Error("invalid command: summarize must be a boolean");
      if (value.includeMessages !== undefined && typeof value.includeMessages !== "boolean") throw new Error("invalid command: includeMessages must be a boolean");
      return { ...base, type, messageId: value.messageId as number, ...(value.summarize === undefined ? {} : { summarize: value.summarize }), ...(value.includeMessages === undefined ? {} : { includeMessages: value.includeMessages }) };
    }
    case "session_fork": {
      if (value.includeMessages !== undefined && typeof value.includeMessages !== "boolean") throw new Error("invalid command: includeMessages must be a boolean");
      return { ...base, type, ...(value.sessionId === undefined ? {} : { sessionId: nonEmpty(value.sessionId, "sessionId") }), ...(value.includeMessages === undefined ? {} : { includeMessages: value.includeMessages }) };
    }
    case "session_recent_messages": return { ...base, type, ...(value.sessionId === undefined ? {} : { sessionId: nonEmpty(value.sessionId, "sessionId") }), ...(positiveLimit(value.limit) === undefined ? {} : { limit: positiveLimit(value.limit)! }) };
    case "session_transcript": return {
      ...base,
      type,
      sessionId: nonEmpty(value.sessionId, "sessionId"),
      ...(nonNegativeCursor(value.cursor) === undefined ? {} : { cursor: nonNegativeCursor(value.cursor)! }),
      ...(positiveLimit(value.limit) === undefined ? {} : { limit: positiveLimit(value.limit)! }),
    };
    case "config_update": return { ...base, type, patch: patch(value.patch) };
    case "mcp_upsert": {
      const envRaw = value.env === undefined ? undefined : patch(value.env, "env");
      if (envRaw !== undefined && Object.values(envRaw).some(item => typeof item !== "string" && item !== null)) throw new Error("invalid command: env values must be strings or null");
      return { ...base, type, name: nonEmpty(value.name, "name"), command: nonEmpty(value.command, "command"), ...(value.args === undefined ? {} : { args: stringArray(value.args, "args") }), ...(envRaw === undefined ? {} : { env: envRaw as Record<string, string | null> }) };
    }
    case "mcp_remove": return { ...base, type, name: nonEmpty(value.name, "name") };
    case "cron_list": {
      if (value.editableOnly !== undefined && typeof value.editableOnly !== "boolean") throw new Error("invalid command: editableOnly must be a boolean");
      const cursor = nonNegativeCursor(value.cursor);
      const limit = positiveLimit(value.limit);
      const maxPromptLength = positiveInteger(value.maxPromptLength, "maxPromptLength");
      return {
        ...base,
        type,
        ...(cursor === undefined ? {} : { cursor }),
        ...(limit === undefined ? {} : { limit }),
        ...(value.editableOnly === undefined ? {} : { editableOnly: value.editableOnly }),
        ...(maxPromptLength === undefined ? {} : { maxPromptLength }),
      };
    }
    case "cron_add": {
      if (value.includeJob !== undefined && typeof value.includeJob !== "boolean") throw new Error("invalid command: includeJob must be a boolean");
      return { ...base, type, schedule: nonEmpty(value.schedule, "schedule"), prompt: nonEmpty(value.prompt, "prompt"), ...(value.jobId === undefined ? {} : { jobId: nonEmpty(value.jobId, "jobId") }), ...(value.includeJob === undefined ? {} : { includeJob: value.includeJob }) };
    }
    case "cron_update": {
      if (value.includeJob !== undefined && typeof value.includeJob !== "boolean") throw new Error("invalid command: includeJob must be a boolean");
      return { ...base, type, jobId: nonEmpty(value.jobId, "jobId"), patch: patch(value.patch) as { schedule?: string; prompt?: string }, ...(value.includeJob === undefined ? {} : { includeJob: value.includeJob }) };
    }
    case "cron_remove": return { ...base, type, jobId: nonEmpty(value.jobId, "jobId") };
    case "memory_list": return { ...base, type, ...(positiveLimit(value.limit) === undefined ? {} : { limit: positiveLimit(value.limit)! }) };
    case "memory_search": return { ...base, type, query: nonEmpty(value.query, "query"), ...(positiveLimit(value.limit) === undefined ? {} : { limit: positiveLimit(value.limit)! }) };
    case "memory_create": return { ...base, type, content: nonEmpty(value.content, "content"), category: nonEmpty(value.category, "category") };
    case "memory_update": return { ...base, type, memoryId: nonEmpty(value.memoryId, "memoryId"), patch: patch(value.patch) as { content?: string; category?: string } };
    case "memory_delete": return { ...base, type, memoryId: nonEmpty(value.memoryId, "memoryId") };
    case "notes_import": {
      if (value.semantic !== undefined && typeof value.semantic !== "boolean") throw new Error("invalid command: semantic must be a boolean");
      return { ...base, type, folderPath: nonEmpty(value.folderPath, "folderPath"), ...(value.semantic === undefined ? {} : { semantic: value.semantic }) };
    }
    case "notes_search": {
      if (value.mode !== undefined && value.mode !== "keyword" && value.mode !== "semantic") throw new Error("invalid command: mode must be keyword or semantic");
      return { ...base, type, query: nonEmpty(value.query, "query"), ...(value.mode === undefined ? {} : { mode: value.mode }), ...(positiveLimit(value.limit) === undefined ? {} : { limit: positiveLimit(value.limit)! }) };
    }
    case "instruction_file_get": return { ...base, type, fileId: nonEmpty(value.fileId, "fileId") };
    case "instruction_file_update": {
      if (typeof value.content !== "string") throw new Error("invalid command: content must be a string");
      return { ...base, type, fileId: nonEmpty(value.fileId, "fileId"), content: value.content };
    }
    case "skill_get": return { ...base, type, name: nonEmpty(value.name, "name") };
    default: throw new Error(`unknown command: ${type}`);
  }
};

File: crates/railgun-backend/src/protocol.rs
use anyhow::{Result, bail};
use serde_json::{Map, Value};

pub const VERSION: i64 = 1;
pub const CAPABILITIES: &[&str] = &[
    "sessions",
    "interaction.approval",
    "interaction.clarification",
    "config",
    "mcp",
    "cron",
    "memory",
    "dream",
    "instructions",
    "skills",
    "session.delivery",
    "model_catalog.refresh",
];

#[derive(Clone, Debug)]
pub struct Command {
    pub id: Option<String>,
    pub kind: String,
    pub fields: Map<String, Value>,
}

impl Command {
    pub fn parse(value: Value) -> Result<Self> {
        let Value::Object(fields) = value else {
            bail!("invalid command: missing type field");
        };
        let kind = fields
            .get("type")
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty())
            .ok_or_else(|| anyhow::anyhow!("invalid command: missing type field"))?
            .to_owned();
        let id = match fields.get("id") {
            None => None,
            Some(value) => Some(non_empty(value, "id")?.to_owned()),
        };
        validate(&kind, &fields)?;
        Ok(Self { id, kind, fields })
    }

    pub fn string(&self, field: &str) -> Result<&str> {
        non_empty(
            self.fields.get(field).ok_or_else(|| {
                anyhow::anyhow!("invalid command: {field} must be a non-empty string")
            })?,
            field,
        )
    }

    pub fn bool(&self, field: &str) -> Result<bool> {
        self.fields
            .get(field)
            .and_then(Value::as_bool)
            .ok_or_else(|| anyhow::anyhow!("invalid command: {field} must be a boolean"))
    }

    pub fn integer(&self, field: &str) -> Result<i64> {
        self.fields
            .get(field)
            .and_then(Value::as_i64)
            .ok_or_else(|| anyhow::anyhow!("invalid command: {field} must be an integer"))
    }
}

fn non_empty<'a>(value: &'a Value, field: &str) -> Result<&'a str> {
    value
        .as_str()
        .filter(|value| !value.trim().is_empty())
        .ok_or_else(|| anyhow::anyhow!("invalid command: {field} must be a non-empty string"))
}

fn optional_bool(fields: &Map<String, Value>, field: &str) -> Result<()> {
    if fields.get(field).is_some_and(|value| !value.is_boolean()) {
        bail!("invalid command: {field} must be a boolean");
    }
    Ok(())
}

fn optional_limit(fields: &Map<String, Value>) -> Result<()> {
    if let Some(value) = fields.get("limit") {
        let valid = value
            .as_i64()
            .is_some_and(|value| (1..=100).contains(&value));
        if !valid {
            bail!("invalid command: limit must be an integer between 1 and 100");
        }
    }
    Ok(())
}

fn optional_cursor(fields: &Map<String, Value>) -> Result<()> {
    if fields
        .get("cursor")
        .is_some_and(|value| value.as_i64().is_none_or(|value| value < 0))
    {
        bail!("invalid command: cursor must be a non-negative integer");
    }
    Ok(())
}

fn require_string(fields: &Map<String, Value>, field: &str) -> Result<()> {
    non_empty(
        fields.get(field).ok_or_else(|| {
            anyhow::anyhow!("invalid command: {field} must be a non-empty string")
        })?,
        field,
    )?;
    Ok(())
}

fn require_text(fields: &Map<String, Value>, field: &str) -> Result<()> {
    if !fields.get(field).is_some_and(Value::is_string) {
        bail!("invalid command: {field} must be a string");
    }
    Ok(())
}

fn require_object(fields: &Map<String, Value>, field: &str) -> Result<()> {
    if !fields.get(field).is_some_and(Value::is_object) {
        bail!("invalid command: {field} must be an object");
    }
    Ok(())
}

fn validate(kind: &str, fields: &Map<String, Value>) -> Result<()> {
    match kind {
        "initialize" => {
            if fields.get("version").and_then(Value::as_i64).is_none() {
                bail!("invalid command: version must be an integer");
            }
            if fields
                .get("clientName")
                .is_some_and(|value| !value.is_string())
            {
                bail!("invalid command: clientName must be a string");
            }
        }
        "prompt" | "steer" | "follow_up" => require_string(fields, "message")?,
        "set_model" => require_string(fields, "modelId")?,
        "set_auto_compaction" => {
            if !fields.get("enabled").is_some_and(Value::is_boolean) {
                bail!("invalid command: enabled must be a boolean");
            }
        }
        "approval_response" => {
            require_string(fields, "requestId")?;
            if !fields.get("approved").is_some_and(Value::is_boolean) {
                bail!("invalid command: approved must be a boolean");
            }
        }
        "clarification_response" => {
            require_string(fields, "requestId")?;
            require_string(fields, "answer")?;
        }
        "session_new" => {
            if fields.contains_key("modelId") {
                require_string(fields, "modelId")?;
            }
        }
        "session_load" => {
            require_string(fields, "sessionId")?;
            optional_bool(fields, "includeMessages")?;
        }
        "session_archive" | "session_unarchive" => require_string(fields, "sessionId")?,
        "session_branch" => {
            if fields
                .get("messageId")
                .and_then(Value::as_i64)
                .is_none_or(|value| value < 1)
            {
                bail!("invalid command: messageId must be a positive integer");
            }
            optional_bool(fields, "summarize")?;
            optional_bool(fields, "includeMessages")?;
        }
        "session_fork" => {
            if fields.contains_key("sessionId") {
                require_string(fields, "sessionId")?;
            }
            optional_bool(fields, "includeMessages")?;
        }
        "session_recent_messages" => {
            if fields.contains_key("sessionId") {
                require_string(fields, "sessionId")?;
            }
            optional_limit(fields)?;
        }
        "session_transcript" => {
            require_string(fields, "sessionId")?;
            optional_cursor(fields)?;
            optional_limit(fields)?;
        }
        "config_update" => require_object(fields, "patch")?,
        "mcp_upsert" => {
            require_string(fields, "name")?;
            require_string(fields, "command")?;
            if let Some(args) = fields.get("args") {
                let valid = args
                    .as_array()
                    .is_some_and(|values| values.iter().all(Value::is_string));
                if !valid {
                    bail!("invalid command: args must be an array of strings");
                }
            }
            if let Some(env) = fields.get("env") {
                let valid = env.as_object().is_some_and(|values| {
                    values
                        .values()
                        .all(|value| value.is_string() || value.is_null())
                });
                if !valid {
                    bail!("invalid command: env values must be strings or null");
                }
            }
        }
        "mcp_remove" => require_string(fields, "name")?,
        "cron_list" => {
            optional_cursor(fields)?;
            optional_limit(fields)?;
            optional_bool(fields, "editableOnly")?;
            if fields
                .get("maxPromptLength")
                .is_some_and(|value| value.as_i64().is_none_or(|value| value < 1))
            {
                bail!("invalid command: maxPromptLength must be a positive integer");
            }
        }
        "cron_add" => {
            require_string(fields, "schedule")?;
            require_string(fields, "prompt")?;
            if fields.contains_key("jobId") {
                require_string(fields, "jobId")?;
            }
            optional_bool(fields, "includeJob")?;
        }
        "cron_update" => {
            require_string(fields, "jobId")?;
            require_object(fields, "patch")?;
            optional_bool(fields, "includeJob")?;
        }
        "cron_remove" => require_string(fields, "jobId")?,
        "memory_list" => optional_limit(fields)?,
        "memory_search" => {
            require_string(fields, "query")?;
            optional_limit(fields)?;
        }
        "memory_create" => {
            require_string(fields, "content")?;
            require_string(fields, "category")?;
        }
        "memory_update" => {
            require_string(fields, "memoryId")?;
            require_object(fields, "patch")?;
        }
        "memory_delete" => require_string(fields, "memoryId")?,
        "instruction_file_get" => require_string(fields, "fileId")?,
        "instruction_file_update" => {
            require_string(fields, "fileId")?;
            if !fields.get("content").is_some_and(Value::is_string) {
                bail!("invalid command: content must be a string");
            }
        }
        "skill_get" | "skill_delete" => require_string(fields, "name")?,
        "skill_create" | "skill_update" => {
            require_string(fields, "name")?;
            require_string(fields, "description")?;
            require_text(fields, "body")?;
            optional_bool(fields, "disableModelInvocation")?;
        }
        "abort"
        | "get_state"
        | "get_messages"
        | "get_available_models"
        | "refresh_model_catalog"
        | "compact"
        | "session_list"
        | "session_list_archived"
        | "session_save"
        | "session_delivery_cursor"
        | "config_get"
        | "mcp_list"
        | "skills_list"
        | "dream_run"
        | "instruction_files_list" => {}
        _ => bail!("unknown command: {kind}"),
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn validates_limits_and_ids() {
        assert_eq!(
            Command::parse(json!({"type":"session_transcript","sessionId":"s","limit":101}))
                .unwrap_err()
                .to_string(),
            "invalid command: limit must be an integer between 1 and 100"
        );
        assert!(Command::parse(json!({"id":"1","type":"initialize","version":1})).is_ok());
    }

    #[test]
    fn accepts_the_additive_fieldless_catalog_refresh_command() {
        assert!(Command::parse(json!({"id":"refresh-1","type":"refresh_model_catalog"})).is_ok());
        assert!(CAPABILITIES.contains(&"model_catalog.refresh"));
    }

    #[test]
    fn validates_skill_crud_shapes_without_rejecting_an_empty_body() {
        assert!(
            Command::parse(json!({
                "type":"skill_create",
                "name":"review",
                "description":"Review the change",
                "body":"",
                "disableModelInvocation":false
            }))
            .is_ok()
        );
        assert!(
            Command::parse(json!({
                "type":"skill_update",
                "name":"review",
                "description":"Review the change",
                "body":"body"
            }))
            .is_ok()
        );
        assert!(Command::parse(json!({"type":"skill_delete","name":"review"})).is_ok());
        assert!(
            Command::parse(json!({
                "type":"skill_create",
                "name":"review",
                "description":"Review"
            }))
            .is_err()
        );
        assert!(
            Command::parse(json!({
                "type":"skill_create",
                "name":"review",
                "description":"Review",
                "body":"body",
                "disableModelInvocation":"false"
            }))
            .is_err()
        );
    }
}
